chacha20poly1305/src/lib.rs

```diff
@@ -0,0 +1,157 @@
+//! ChaCha20Poly1305 Authenticated Encryption with Additional Data Algorithm
+//! (RFC 8439)
+
+#![no_std]
+
+extern crate alloc;
+
+pub use aead;
+
+use aead::generic_array::typenum::{U0, U12, U16, U32};
+use aead::{generic_array::GenericArray, Aead, Error, NewAead};
+use alloc::vec::Vec;
+use chacha20::stream_cipher::{NewStreamCipher, SyncStreamCipher, SyncStreamCipherSeek};
+use chacha20::ChaCha20;
+use core::convert::TryInto;
+use poly1305::{Poly1305, Tag};
+use zeroize::{Zeroize, Zeroizing};
+
+/// ChaCha20Poly1305 AEAD
+#[derive(Clone)]
+pub struct ChaCha20Poly1305 {
+    /// Secret key
+    key: GenericArray<u8, U32>,
+}
+
+impl NewAead for ChaCha20Poly1305 {
+    type KeySize = U32;
+
+    fn new(key: GenericArray<u8, U32>) -> Self {
+        ChaCha20Poly1305 { key }
+    }
+}
+
+impl Aead for ChaCha20Poly1305 {
+    type NonceSize = U12;
+    type TagSize = U16;
+    type CiphertextOverhead = U0;
+
+    fn encrypt(
+        &mut self,
+        associated_data: &[u8],
+        nonce: &GenericArray<u8, Self::NonceSize>,
+        plaintext: &[u8],
+    ) -> Result<Vec<u8>, Error> {
+        CipherInstance::new(&self.key, nonce).encrypt(associated_data, plaintext)
+    }
+
+    fn decrypt(
+        &mut self,
+        associated_data: &[u8],
+        nonce: &GenericArray<u8, Self::NonceSize>,
+        ciphertext: &[u8],
+    ) -> Result<Vec<u8>, Error> {
+        CipherInstance::new(&self.key, nonce).decrypt(associated_data, ciphertext)
+    }
+}
+
+impl Drop for ChaCha20Poly1305 {
+    fn drop(&mut self) {
+        self.key.as_mut_slice().zeroize();
+    }
+}
+
+/// ChaCha20Poly1305 instantiated with a particular nonce
+struct CipherInstance {
+    chacha20: ChaCha20,
+    poly1305: Poly1305,
+}
```

Comment:
Looks like it can be made generic over `StreamCipher` trait? I wonder if it should not go to the `poly1305` crate...

Reply:
This is a first cut 😉But yeah, good point.

It would definitely be nice to reuse the same generic "Poly1305 plus a Salsa20 family stream cipher" core to implement Salsa20Poly1305, ChaCha20Poly1305, and XChaCha20Poly1305.

The `poly1305` crate might make sense as it's the eponymous algorithm's primary use case. I could also see this further justifying the existence of `salsa20-core` versus trying to abstract it away into a more generic `ctr`.